src/fsharp/NameResolution.fs

```diff
@@ -770,57 +788,376 @@ let AddUnionCases2 bulkAddMode (eUnqualifiedItems: UnqualifiedItems) (ucrefs: Un
+                        let tcref = tcref.NestedTyconRef entity
+                        if IsEntityAccessible amap m ad tcref then Some tcref else None)
+        | _ -> [])
+
+/// Make a type that refers to a nested type.
+///
+/// Handle the .NET/C# business where nested generic types implicitly accumulate the type parameters
+/// from their enclosing types.
+let MakeNestedType (ncenv: NameResolver) (tinst: TType list) m (tcrefNested: TyconRef) =
+    let tps = List.skip tinst.Length (tcrefNested.Typars m)
+    let tinstNested = ncenv.InstantiationGenerator m tps
+    mkAppTy tcrefNested (tinst @ tinstNested)
+
+/// Get all the accessible nested types of an existing type.
+let GetNestedTypesOfType (ad, ncenv: NameResolver, optFilter, staticResInfo, checkForGenerated, m) ty =
+    let tinst, tcrefsNested = GetNestedTyconRefsOfType ncenv.InfoReader ncenv.amap (ad, optFilter, staticResInfo, checkForGenerated, m) ty
+    tcrefsNested
+    |> List.map (MakeNestedType ncenv tinst m)
+
+let ChooseMethInfosForNameEnv g m ty (minfos: MethInfo list) =
+    let isExtTy = IsTypeUsedForCSharpStyleExtensionMembers g m ty
+
+    let methGroups =
+        minfos
+        |> List.filter (fun minfo ->
+            not (minfo.IsInstance || minfo.IsClassConstructor || minfo.IsConstructor) && typeEquiv g minfo.ApparentEnclosingType ty &&
+            not (IsMethInfoPlainCSharpStyleExtensionMember g m isExtTy minfo))
+        |> List.groupBy (fun minfo -> minfo.LogicalName)
+
+    seq {
+        for (methName, methGroup) in methGroups do
+            if not methGroup.IsEmpty then
+                KeyValuePair(methName, Item.MethodGroup(methName, methGroup, None)) 
+    }
+
+let ChoosePropInfosForNameEnv g ty (pinfos: PropInfo list) =
+    let propGroups =
+        pinfos
+        |> List.filter (fun pinfo ->
+            pinfo.IsStatic && typeEquiv g pinfo.ApparentEnclosingType ty)
+        |> List.groupBy (fun pinfo -> pinfo.PropertyName)
+
+    seq {
+        for (propName, propGroup) in propGroups do
+            if not propGroup.IsEmpty then
+                KeyValuePair(propName, Item.Property(propName, propGroup))
+    }
+
+let ChooseILFieldInfosForNameEnv g ty (finfos: ILFieldInfo list) =
+    seq {
+        for finfo in finfos do
+            if finfo.IsStatic && typeEquiv g finfo.ApparentEnclosingType ty then
+                KeyValuePair(finfo.FieldName, Item.ILField finfo)
+    }
+
+let ChooseEventInfosForNameEnv g ty (einfos: EventInfo list) =
+    seq {
+        for einfo in einfos do
+            if einfo.IsStatic && typeEquiv g einfo.ApparentEnclosingType ty then
+                KeyValuePair(einfo.EventName, Item.Event einfo)
+    }
+
+/// Add content from a type.
+/// Rules:
+///     1. Add nested types.
+///     2. Add C# style extension members.
+///     3. Add extention methods.
+///     4. Add extension properties.
+///     5. Add events.
+///     6. Add fields.
+///     7. Add properies.
+///     8. Add methods.
```

Comment:
Note: my understanding from touching this area before is that the order didn't matter then, so it likely doesn't now. But I'm being extra pendantic here since this kind of comment tends to strongly inform how I interpret the meaning of the function that follows, and when it's laid out this way it implies an ordering.

Reply:
I didn't decide this order. The order matters if it's to be consistent to what gets resolved after dotting into a type, which this was the best way I could think of.

This is unfortunately confusing as no one EVER thinks about it but it does matter.

Example:
```fsharp
type Test = class end

module TextExtensions =

    type Test with
        static member Prop = 1

module TestExtensions2 =

    type Test with
        static member Prop() = "a string"

open TestExtensions
open TestExtensions2

let x : string = Test.Prop() // error - 'Prop' defined in the first TestExtensions takes priority
let y : int = Test.Prop // succeeds

open type Test

let xx : string = Prop() // error - 'Prop' defined in the first TestExtensions takes priority - We need to mimic the behavior above and we do that by specifying the order when adding the content
let yy : int = Prop // succeeds
```